node/pod.go

```diff
@@ -226,78 +226,104 @@ func (pc *PodController) updatePodStatuses(ctx context.Context, q workqueue.Rate
 
 	for _, pod := range pods {
 		if !shouldSkipPodStatusUpdate(pod) {
-			enqueuePodStatusUpdate(ctx, q, pod)
+			pc.fetchPodStatusFromProvider(ctx, q, pod)
 		}
 	}
 }
+func (pc *PodController) fetchPodStatusFromProvider(ctx context.Context, q workqueue.RateLimitingInterface, podFromKubernetes *corev1.Pod) {
+	podStatus, err := pc.provider.GetPodStatus(ctx, podFromKubernetes.Namespace, podFromKubernetes.Name)
+	if errdefs.IsNotFound(err) || (err == nil && podStatus == nil) {
+		// Only change the status when the pod was already up
+		// Only doing so when the pod was successfully running makes sure we don't run into race conditions during pod creation.
+		if podFromKubernetes.Status.Phase == corev1.PodRunning || time.Since(podFromKubernetes.ObjectMeta.CreationTimestamp.Time) > time.Minute {
+			// Set the pod to failed, this makes sure if the underlying container implementation is gone that a new pod will be created.
+			podStatus = podFromKubernetes.Status.DeepCopy()
+			podStatus.Phase = corev1.PodFailed
+			podStatus.Reason = "NotFound"
+			podStatus.Message = "The pod status was not found and may have been deleted from the provider"
+			now := metav1.NewTime(time.Now())
+			for i, c := range podStatus.ContainerStatuses {
+				if c.State.Running == nil {
+					continue
+				}
+				podStatus.ContainerStatuses[i].State.Terminated = &corev1.ContainerStateTerminated{
```

Comment:
So if a provider sets a pod's `ContainerStatuses`, VK is overwriting it? I know this hasn't changed behavior in `master` but I'm wondering if we overlooked this in the past.

Reply:
Correct. This is a lift-and-shift.

I think @chewong has some work in this area he's doing. 

Long-term, I suggest we do not change this code, because it will / should go away.

Reply:
Maybe just add a TODO?

Reply:
Wait, thinking back, should we change this behaviour? If a container is suddenly not found, should we change that? 

Do you want to file an issue discussing your concern, and we can figure out what to do from there?

Reply:
Yes, I'll start an issue.